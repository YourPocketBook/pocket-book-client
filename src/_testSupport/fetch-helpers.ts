import fetchMock, {
  MockOptionsMethodDelete,
  MockOptionsMethodGet,
  MockOptionsMethodPost,
  MockOptionsMethodPut,
  MockResponse,
  MockResponseFunction
} from "fetch-mock";

afterEach(fetchMock.reset);

export function authenticatedGet(
  url: string,
  token: string,
  response: MockResponse | MockResponseFunction,
  options?: MockOptionsMethodGet
) {
  fetchMock.get(
    (u, h) =>
      u === url &&
      (h.headers as { Authorization: string }).Authorization ===
        `Bearer ${token}`,
    response,
    options
  );
}

export function authenticatedPut(
  url: string,
  token: string,
  body: {},
  response: MockResponse | MockResponseFunction,
  options?: MockOptionsMethodPut
) {
  fetchMock.put(
    (u, h) =>
      u === url &&
      (h.headers as { Authorization: string }).Authorization ===
        `Bearer ${token}` &&
      h.body === JSON.stringify(body),
    response,
    options
  );
}

export function authenticatedPost(
  url: string,
  token: string,
  body: {},
  response: MockResponse | MockResponseFunction,
  options?: MockOptionsMethodPost
) {
  fetchMock.post(
    (u, h) =>
      u === url &&
      (h.headers as { Authorization: string }).Authorization ===
        `Bearer ${token}` &&
      h.body === JSON.stringify(body),
    response,
    options
  );
}

export function authenticatedDelete(
  url: string,
  token: string,
  response: MockResponse | MockResponseFunction,
  options?: MockOptionsMethodDelete
) {
  fetchMock.delete(
    (u, h) =>
      u === url &&
      (h.headers as { Authorization: string }).Authorization ===
        `Bearer ${token}`,
    response,
    options
  );
}
