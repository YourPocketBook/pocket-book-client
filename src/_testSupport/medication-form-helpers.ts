import { fireEvent } from "@testing-library/react";

export const testDefaultState = {
  adviceIfDeclined: "Test Advice 1",
  adviceIfTaken: "Test Advice 2",
  dose: "Test Dose",
  exclusionCriteria: "Test Exclusion",
  form: "Test Form",
  inclusionCriteria: "Test Inclusion",
  indications: "Test Indications",
  name: "Test Name",
  policyDate: "2019-12-10",
  route: "Test Route",
  sideEffects: "Test Side Effects"
};

export function setupState(getByLabelText: (text: string) => HTMLElement) {
  fireEvent.change(getByLabelText("Name"), {
    target: { value: testDefaultState.name }
  });
  fireEvent.change(getByLabelText("Policy Date"), {
    target: { value: testDefaultState.policyDate }
  });
  fireEvent.change(getByLabelText("Indications"), {
    target: { value: testDefaultState.indications }
  });
  fireEvent.change(getByLabelText("Inclusion Criteria"), {
    target: { value: testDefaultState.inclusionCriteria }
  });
  fireEvent.change(getByLabelText("Exclusion Criteria"), {
    target: { value: testDefaultState.exclusionCriteria }
  });
  fireEvent.change(getByLabelText("Name, Forms and Strengths of Medicine"), {
    target: { value: testDefaultState.form }
  });
  fireEvent.change(getByLabelText("Route or Method"), {
    target: { value: testDefaultState.route }
  });
  fireEvent.change(getByLabelText("Dosage"), {
    target: { value: testDefaultState.dose }
  });
  fireEvent.change(getByLabelText("Side Effects"), {
    target: { value: testDefaultState.sideEffects }
  });
  fireEvent.change(getByLabelText("Advice if Taken"), {
    target: { value: testDefaultState.adviceIfTaken }
  });
  fireEvent.change(getByLabelText("Advice if Declined"), {
    target: { value: testDefaultState.adviceIfDeclined }
  });
}

export function checkForClass(
  getByLabelText: (text: string) => HTMLElement,
  cls: "is-valid" | "is-invalid"
) {
  expect(getByLabelText("Name")).toHaveClass(cls);
  expect(getByLabelText("Policy Date")).toHaveClass(cls);
  expect(getByLabelText("Indications")).toHaveClass(cls);
  expect(getByLabelText("Inclusion Criteria")).toHaveClass(cls);
  expect(getByLabelText("Exclusion Criteria")).toHaveClass(cls);
  expect(getByLabelText("Name, Forms and Strengths of Medicine")).toHaveClass(
    cls
  );
  expect(getByLabelText("Route or Method")).toHaveClass(cls);
  expect(getByLabelText("Dosage")).toHaveClass(cls);
  expect(getByLabelText("Side Effects")).toHaveClass(cls);
  expect(getByLabelText("Advice if Taken")).toHaveClass(cls);
  expect(getByLabelText("Advice if Declined")).toHaveClass(cls);
}

export function checkState(
  getByLabelText: (text: string) => HTMLElement,
  hasValue: boolean
) {
  expect(getByLabelText("Name")).toHaveValue(
    hasValue ? testDefaultState.name : ""
  );
  expect(getByLabelText("Policy Date")).toHaveValue(
    hasValue ? testDefaultState.policyDate : ""
  );
  expect(getByLabelText("Indications")).toHaveTextContent(
    hasValue ? testDefaultState.indications : ""
  );
  expect(getByLabelText("Inclusion Criteria")).toHaveTextContent(
    hasValue ? testDefaultState.inclusionCriteria : ""
  );
  expect(getByLabelText("Exclusion Criteria")).toHaveTextContent(
    hasValue ? testDefaultState.exclusionCriteria : ""
  );
  expect(
    getByLabelText("Name, Forms and Strengths of Medicine")
  ).toHaveTextContent(hasValue ? testDefaultState.form : "");
  expect(getByLabelText("Route or Method")).toHaveTextContent(
    hasValue ? testDefaultState.route : ""
  );
  expect(getByLabelText("Dosage")).toHaveTextContent(
    hasValue ? testDefaultState.dose : ""
  );
  expect(getByLabelText("Side Effects")).toHaveTextContent(
    hasValue ? testDefaultState.sideEffects : ""
  );
  expect(getByLabelText("Advice if Taken")).toHaveTextContent(
    hasValue ? testDefaultState.adviceIfTaken : ""
  );
  expect(getByLabelText("Advice if Declined")).toHaveTextContent(
    hasValue ? testDefaultState.adviceIfDeclined : ""
  );
}
