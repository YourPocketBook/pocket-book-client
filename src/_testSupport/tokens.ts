export const inDateNonAdminToken =
  "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwiaWF0IjoxNDE2MjM5MDIyLCJleHAiOjE3MTYyNDAwMjIsImF1ZCI6ImxvY2FsaG9zdCIsImlzcyI6ImxvY2FsaG9zdCIsImdpdmVuX25hbWUiOiJUZXN0IFVzZXIiLCJuYW1laWQiOiJmZjFkMDhlNy00NjU2LTRjMjgtOTc2Zi0xYzI3ZjE5MTQ3N2YiLCJlbWFpbCI6InRlc3RAc2phLm9yZy51ayJ9.VmT5YkTG0sUd5sK7skfMdi9kc5XG7Yhx6sqhqMSxhVY";
export const inDateAdminToken =
  "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwiaWF0IjoxNDE2MjM5MDIyLCJleHAiOjE3MTYyNDAwMjIsImF1ZCI6ImxvY2FsaG9zdCIsImlzcyI6ImxvY2FsaG9zdCIsInJvbGUiOiJBZG1pbiIsImdpdmVuX25hbWUiOiJUZXN0IFVzZXIiLCJuYW1laWQiOiJmZjFkMDhlNy00NjU2LTRjMjgtOTc2Zi0xYzI3ZjE5MTQ3N2YiLCJlbWFpbCI6InRlc3RAc2phLm9yZy51ayJ9.rSQb2Hv6RfxaXSfO6s3QjLjnYKeqVgps3rPenquh3Nk";
export const expiredToken =
  "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwiaWF0IjoxNDE2MjM5MDIyLCJleHAiOjE0MTYyNDAwMjIsImF1ZCI6ImxvY2FsaG9zdCIsImlzcyI6ImxvY2FsaG9zdCIsInVuaXF1ZV9uYW1lIjoiVGVzdCBVc2VyIiwibmFtZWlkIjoiZmYxZDA4ZTctNDY1Ni00YzI4LTk3NmYtMWMyN2YxOTE0NzdmIiwiZW1haWwiOiJ0ZXN0QHNqYS5vcmcudWsifQ.7bDfHDk9sWXL8boFeZWIdo_GB-4Qj8PvA4HDC49EzNc";
