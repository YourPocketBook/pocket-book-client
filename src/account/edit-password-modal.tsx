import React, { useState } from "react";
import Button from "reactstrap/lib/Button";
import Col from "reactstrap/lib/Col";
import Form from "reactstrap/lib/Form";
import FormFeedback from "reactstrap/lib/FormFeedback";
import Modal from "reactstrap/lib/Modal";
import ModalBody from "reactstrap/lib/ModalBody";
import ModalFooter from "reactstrap/lib/ModalFooter";
import ModalHeader from "reactstrap/lib/ModalHeader";
import Row from "reactstrap/lib/Row";
import { Field } from "../field";
import { useSubmitData } from "../hooks/useSubmitData";
import { useValidation } from "../hooks/useValidation";
import { PasswordScore } from "../password-score";

const endPoint = "/api/update-password";

interface IEditPasswordModalProps {
  show: boolean;
  setShow(val: boolean): void;
  done(): void;
}

export function EditPasswordModal(props: IEditPasswordModalProps) {
  const { done, show, setShow } = props;

  const [
    currentPassword,
    currentPasswordValid,
    setCurrentPassword,
    setCurrentPasswordValid
  ] = useValidation("", "text");
  const [
    newPassword,
    newPasswordValid,
    setNewPassword,
    setNewPasswordValid
  ] = useValidation("", "text");
  const [showPassword, setShowPassword] = useState(false);
  const [showPasswordSuggestions, setShowPasswordSuggestions] = useState(false);

  const [showValidation, setShowValidation] = useState(false);

  const { serverError, inProgress, submit } = useSubmitData(
    { currentPassword, newPassword },
    { currentPassword: currentPasswordValid, newPassword: newPasswordValid },
    {
      currentPassword: setCurrentPasswordValid,
      newPassword: setNewPasswordValid
    },
    endPoint,
    {
      authenticate: true,
      method: "PUT",
      setShowValidation,
      success: () => {
        setCurrentPassword("");
        setNewPassword("");
        setShowPassword(false);
        setShow(false);
        done();
      }
    }
  );

  const serverWarning =
    showValidation && serverError ? (
      <FormFeedback className="d-block">
        The server has had an issue, please try again.
      </FormFeedback>
    ) : null;

  return (
    <Modal isOpen={show} toggle={() => setShow(!show)}>
      <ModalHeader>Update your password</ModalHeader>
      <ModalBody>
        <Form onSubmit={submit}>
          {serverWarning}
          <Field
            placeholder="Current Password"
            id="currentPassword"
            name="currentPassword"
            longText={false}
            showValidation={showValidation}
            valid={currentPasswordValid}
            value={currentPassword}
            warning="You need to enter your current password."
            onChange={setCurrentPassword}
            autoComplete="current-password"
            type={showPassword ? "text" : "password"}
          />
          <Field
            placeholder="New Password"
            id="newPassword"
            name="newPassword"
            longText={false}
            showValidation={showValidation}
            valid={newPasswordValid}
            value={newPassword}
            warning="You need to enter a new password."
            onChange={s => {
              setNewPassword(s);
              setShowPasswordSuggestions(true);
            }}
            autoComplete="new-password"
            type={showPassword ? "text" : "password"}
          />
          <Row>
            <Col md={{ size: 9, offset: 3 }} className="text-right">
              <Button
                color="link"
                id="showPassword"
                type="button"
                onClick={() => setShowPassword(!showPassword)}
                className="text-muted"
              >
                {showPassword ? "Hide Password" : "Show Password"}
              </Button>
            </Col>
          </Row>
          <PasswordScore
            password={newPassword}
            show={showPasswordSuggestions}
          />
        </Form>
      </ModalBody>
      <ModalFooter>
        <Button
          color="dark-green"
          disabled={inProgress}
          // tslint:disable-next-line:no-empty
          onClick={() => submit({ preventDefault: () => {} })}
        >
          Save
        </Button>
        <Button
          color="secondary"
          onClick={() => {
            setShow(false);
            setCurrentPassword("");
            setNewPassword("");
            setShowPassword(false);
          }}
        >
          Cancel
        </Button>
      </ModalFooter>
    </Modal>
  );
}
