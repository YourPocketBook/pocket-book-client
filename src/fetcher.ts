import isFuture from "date-fns/isFuture";
import jwtDecode from "jwt-decode";

class Fetcher {
  private realName: string = "";
  private nameChanged?: (name: string) => void;
  private token: string | null;
  private parsedToken:
    | { exp: number; given_name: string; role: string }
    | undefined;

  constructor() {
    this.token = localStorage.getItem("token");
    this.parseToken();
  }

  public setNameChangeHandler(func?: (name: string) => void) {
    this.nameChanged = func;
  }

  public authFetch(input: RequestInfo, init?: RequestInit): Promise<Response> {
    if (!this.token) {
      throw new Error("No token");
    }

    init = {
      ...init,
      headers: {
        ...(init ? init.headers : {}),
        Authorization: `Bearer ${this.token}`
      }
    };

    return fetch(input, init);
  }

  public saveToken(token: string) {
    localStorage.setItem("token", token);
    this.token = token;
    this.parseToken();
  }

  public checkToken(token: string): boolean {
    return this.token === token;
  }

  public clearToken() {
    localStorage.removeItem("token");
    this.token = null;
    this.parsedToken = undefined;
  }

  public getName() {
    return this.realName;
  }

  public updateName(name: string) {
    this.realName = name;
    if (this.nameChanged) {
      this.nameChanged(name);
    }
  }

  public isAdmin() {
    if (this.parsedToken) {
      return this.parsedToken.role === "Admin";
    }
    return false;
  }

  public hasToken(): boolean {
    return !!this.token;
  }

  public isInDate() {
    if (this.parsedToken) {
      return isFuture(this.parsedToken.exp * 1000);
    }

    return false;
  }

  private parseToken() {
    if (this.token) {
      try {
        this.parsedToken = jwtDecode(this.token);
      } catch {
        this.token = null;
        this.updateName("");
        localStorage.removeItem("token");
        this.parsedToken = undefined;
      }

      if (this.parsedToken) {
        this.updateName(this.parsedToken.given_name);
      }
    } else {
      this.parsedToken = undefined;
      this.updateName("");
    }
  }
}

const fetcher = new Fetcher();

export { fetcher, Fetcher };
