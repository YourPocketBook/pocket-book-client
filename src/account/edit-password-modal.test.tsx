import "@testing-library/jest-dom/extend-expect";
import { fireEvent, render, wait } from "@testing-library/react";
import fetchMock from "fetch-mock";
import React from "react";
import ReactDOM from "react-dom";
import { authenticatedPut } from "../_testSupport/fetch-helpers";
import { inDateNonAdminToken } from "../_testSupport/tokens";
import { fetcher } from "../fetcher";
import { history } from "../history";
import { EditPasswordModal } from "./edit-password-modal";

const setPasswordEndPoint = "/api/update-password";

describe("Edit Password Modal", () => {
  beforeAll(() => {
    fetcher.saveToken(inDateNonAdminToken);
  });

  beforeEach(() => {
    history.push = jest.fn();
  });

  it("renders without crashing", () => {
    const div = document.createElement("div");

    ReactDOM.render(
      <EditPasswordModal done={jest.fn()} setShow={jest.fn()} show={true} />,
      div
    );
    ReactDOM.unmountComponentAtNode(div);
  });

  it("renders hidden correctly", () => {
    const { baseElement } = render(
      <EditPasswordModal done={jest.fn()} setShow={jest.fn()} show={false} />
    );

    expect(baseElement).toMatchSnapshot();
  });

  it("renders shown correctly", async () => {
    const { baseElement } = render(
      <EditPasswordModal done={jest.fn()} setShow={jest.fn()} show={true} />
    );

    expect(baseElement).toMatchSnapshot();
  });

  it("hides dialog when cancel button is pressed", () => {
    const showFn = jest.fn();

    const { getByText } = render(
      <EditPasswordModal done={jest.fn()} setShow={showFn} show={true} />
    );

    fireEvent.click(getByText("Cancel"));

    expect(showFn).toBeCalledWith(false);
  });

  it("defaults to correct values", async () => {
    const { getByLabelText } = render(
      <EditPasswordModal done={jest.fn()} setShow={jest.fn()} show={true} />
    );

    expect(getByLabelText("Current Password")).toHaveAttribute("value", "");
    expect(getByLabelText("New Password")).toHaveAttribute("value", "");
  });

  it("declares valid values correctly", async () => {
    const testCurrent = "Test Current";
    const testNew = "Test New";

    authenticatedPut(
      setPasswordEndPoint,
      inDateNonAdminToken,
      { currentPassword: testCurrent, newPassword: testNew },
      { status: 400, body: JSON.stringify([{ path: "doses", error: "test" }]) }
    );

    const { getByLabelText, getByText } = render(
      <EditPasswordModal done={jest.fn()} setShow={jest.fn()} show={true} />
    );

    fireEvent.change(getByLabelText("Current Password"), {
      target: { value: testCurrent }
    });
    fireEvent.change(getByLabelText("New Password"), {
      target: { value: testNew }
    });

    fireEvent.click(getByText("Save"));

    await wait(() => expect(fetchMock.done()).toBeTruthy());

    expect(getByLabelText("Current Password")).toHaveClass("is-valid");
    expect(getByLabelText("New Password")).toHaveClass("is-valid");
  });

  it("declares invalid values correctly", () => {
    const { getByLabelText, getByText } = render(
      <EditPasswordModal done={jest.fn()} setShow={jest.fn()} show={true} />
    );

    fireEvent.click(getByText("Save"));

    expect(getByLabelText("Current Password")).toHaveClass("is-invalid");
    expect(getByLabelText("New Password")).toHaveClass("is-invalid");
  });

  it("renders server error correctly", async () => {
    const testCurrent = "Test Current";
    const testNew = "Test New";

    authenticatedPut(
      setPasswordEndPoint,
      inDateNonAdminToken,
      { currentPassword: testCurrent, newPassword: testNew },
      { status: 500 }
    );

    const { getByLabelText, getByText } = render(
      <EditPasswordModal done={jest.fn()} setShow={jest.fn()} show={true} />
    );

    fireEvent.change(getByLabelText("Current Password"), {
      target: { value: testCurrent }
    });
    fireEvent.change(getByLabelText("New Password"), {
      target: { value: testNew }
    });
    fireEvent.click(getByText("Save"));

    await wait(() => expect(fetchMock.done()).toBeTruthy());

    expect(getByText(/The server has had an issue/)).toBeInTheDocument();
  });

  it("renders in progress correctly", () => {
    const testCurrent = "Test Current";
    const testNew = "Test New";

    authenticatedPut(
      setPasswordEndPoint,
      inDateNonAdminToken,
      { currentPassword: testCurrent, newPassword: testNew },
      // tslint:disable-next-line: no-empty
      new Promise(() => {})
    );

    const { getByText, getByLabelText } = render(
      <EditPasswordModal done={jest.fn()} setShow={jest.fn()} show={true} />
    );

    fireEvent.change(getByLabelText("Current Password"), {
      target: { value: testCurrent }
    });
    fireEvent.change(getByLabelText("New Password"), {
      target: { value: testNew }
    });

    fireEvent.click(getByText("Save"));
    expect(getByText("Save")).toBeDisabled();
  });

  it("calls done on successful form submit", async () => {
    const testFn = jest.fn();
    const testCurrent = "Test Current";
    const testNew = "Test New";

    authenticatedPut(
      setPasswordEndPoint,
      inDateNonAdminToken,
      { currentPassword: testCurrent, newPassword: testNew },
      { status: 204 }
    );

    const { getByText, getByLabelText } = render(
      <EditPasswordModal done={testFn} setShow={jest.fn()} show={true} />
    );

    fireEvent.change(getByLabelText("Current Password"), {
      target: { value: testCurrent }
    });
    fireEvent.change(getByLabelText("New Password"), {
      target: { value: testNew }
    });
    fireEvent.click(getByText("Save"));
    await wait(() => expect(fetchMock.done()).toBeTruthy());

    expect(testFn).toBeCalled();
  });

  it("toggles show password when button is clicked", () => {
    const { getByText, getByLabelText } = render(
      <EditPasswordModal done={jest.fn()} setShow={jest.fn()} show={true} />
    );

    fireEvent.click(getByText("Show Password"));
    expect(getByLabelText("Current Password")).toHaveAttribute("type", "text");
    expect(getByLabelText("New Password")).toHaveAttribute("type", "text");

    fireEvent.click(getByText("Hide Password"));
    expect(getByLabelText("Current Password")).toHaveAttribute(
      "type",
      "password"
    );
    expect(getByLabelText("New Password")).toHaveAttribute("type", "password");
  });
});
