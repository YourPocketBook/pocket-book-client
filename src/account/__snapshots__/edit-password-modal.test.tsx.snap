// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Edit Password Modal renders hidden correctly 1`] = `
<body
  class=""
>
  <div />
</body>
`;

exports[`Edit Password Modal renders shown correctly 1`] = `
<body
  class="modal-open"
>
  <div />
  <div
    style="position: relative; z-index: 1050; display: block;"
    tabindex="-1"
  >
    <div
      class=""
    >
      <div
        class="modal fade"
        role="dialog"
        style="display: block;"
        tabindex="-1"
      >
        <div
          class="modal-dialog"
          role="document"
        >
          <div
            class="modal-content"
          >
            <div
              class="modal-header"
            >
              <h5
                class="modal-title"
              >
                Update your password
              </h5>
            </div>
            <div
              class="modal-body"
            >
              <form
                class=""
              >
                <div
                  class="row"
                >
                  <div
                    class="col"
                  >
                    <div
                      class="form-group"
                    >
                      <label
                        class=""
                        for="currentPassword"
                      >
                        Current Password
                      </label>
                      <input
                        autocomplete="current-password"
                        class="form-control"
                        id="currentPassword"
                        name="currentPassword"
                        placeholder="Current Password"
                        type="password"
                        value=""
                      />
                      <div
                        class="invalid-feedback"
                      >
                        You need to enter your current password.
                      </div>
                    </div>
                  </div>
                </div>
                <div
                  class="row"
                >
                  <div
                    class="col"
                  >
                    <div
                      class="form-group"
                    >
                      <label
                        class=""
                        for="newPassword"
                      >
                        New Password
                      </label>
                      <input
                        autocomplete="new-password"
                        class="form-control"
                        id="newPassword"
                        name="newPassword"
                        placeholder="New Password"
                        type="password"
                        value=""
                      />
                      <div
                        class="invalid-feedback"
                      >
                        You need to enter a new password.
                      </div>
                    </div>
                  </div>
                </div>
                <div
                  class="row"
                >
                  <div
                    class="text-right col-md-9 offset-md-3"
                  >
                    <button
                      class="text-muted btn btn-link"
                      id="showPassword"
                      type="button"
                    >
                      Show Password
                    </button>
                  </div>
                </div>
              </form>
            </div>
            <div
              class="modal-footer"
            >
              <button
                class="btn btn-dark-green"
                type="button"
              >
                Save
              </button>
              <button
                class="btn btn-secondary"
                type="button"
              >
                Cancel
              </button>
            </div>
          </div>
        </div>
      </div>
      <div
        class="modal-backdrop fade"
      />
    </div>
  </div>
</body>
`;
